libcontainer/mount_linux.go

```diff
@@ -121,3 +179,84 @@ func syscallMode(i fs.FileMode) (o uint32) {
 	// No mapping for Go's ModeTemporary (plan9 only).
 	return
 }
+
+// mountFd creates a "mount source fd" (either through open_tree(2) or just
+// open(O_PATH)) based on the provided configuration. This function must be
+// called from within the container's mount namespace.
+//
+// In the case of idmapped mount configurations, the returned mount source will
+// be an open_tree(2) file with MOUNT_ATTR_IDMAP applied. For other
+// bind-mounts, it will be an O_PATH. If the type of mount cannot be handled,
+// the returned mountSource will be nil, indicating that the container init
+// process will need to do an old-fashioned mount(2) themselves.
+//
+// This helper is only intended to be used by goCreateMountSources.
+func mountFd(nsHandles *userns.Handles, m *configs.Mount) (*mountSource, error) {
+	if !m.IsBind() {
+		return nil, errors.New("new mount api: only bind-mounts are supported")
+	}
+	if nsHandles == nil {
+		nsHandles = new(userns.Handles)
+		defer nsHandles.Release()
+	}
+
+	var mountFile *os.File
+	var sourceType mountSourceType
+
+	if m.IsBind() {
+		// We only need to use OPEN_TREE_CLONE in the case where we need to use
+		// mount_setattr(2). We are currently in the container namespace and
+		// thus there is no risk of an opened directory being used to escape
+		// the container. OPEN_TREE_CLONE is more expensive than open(2)
+		// because it requires doing mounts inside a new anonymous mount
+		// namespace.
+		if m.IsIDMapped() {
+			flags := uint(unix.OPEN_TREE_CLONE | unix.OPEN_TREE_CLOEXEC)
+			if m.Flags&unix.MS_REC == unix.MS_REC {
+				flags |= unix.AT_RECURSIVE
+			}
+			fd, err := unix.OpenTree(unix.AT_FDCWD, m.Source, flags)
+			if err != nil {
+				return nil, &os.PathError{Op: "open_tree(OPEN_TREE_CLONE)", Path: m.Source, Err: err}
+			}
+			mountFile = os.NewFile(uintptr(fd), m.Source)
+			sourceType = mountSourceOpenTree
+
+			// Configure the id mapping.
+			usernsFile, err := nsHandles.Get(userns.Mapping{
+				UIDMappings: m.UIDMappings,
+				GIDMappings: m.GIDMappings,
+			})
+			if err != nil {
+				return nil, fmt.Errorf("failed to create userns for %s id-mapping: %w", m.Source, err)
+			}
+			defer usernsFile.Close()
+
+			setAttrFlags := uint(unix.AT_EMPTY_PATH)
+			// If the mount has "ridmap" set, we apply the configuration
+			// recursively. This allows you to create "rbind" mounts where only
+			// the top-level mount has an idmapping. I'm not sure why you'd
+			// want that, but still...
+			if m.Extensions&configs.EXT_REC_IDMAP != 0 {
```

Comment:
A not important thing:
How about add one more condition check `m.Flags&unix.MS_REC == unix.MS_REC` besides this check?
For example: `"bind", "ridmap"`.

Reply:
`bind,ridmap` is functionally the same as `bind,idmap`, so adding more logic doesn't really accomplish anything and IMHO the logic is much clearer if you only check for `MS_REC` for the `open_tree(2)` call (emulating `rbind`) and only `EXT_REC_IDMAP` for `mount_setattr(2)`, because it's clear that the two flags are not related.